aiida/manage/configuration/migrations/migrations.py

```diff
@@ -8,136 +8,282 @@
 ###########################################################################
 """Define the current configuration version and migrations."""
+from typing import Any, Dict, Protocol
 
-__all__ = ('CURRENT_CONFIG_VERSION', 'OLDEST_COMPATIBLE_CONFIG_VERSION')
+from aiida.common import exceptions
+
+__all__ = (
+    'CURRENT_CONFIG_VERSION', 'OLDEST_COMPATIBLE_CONFIG_VERSION', 'get_current_version', 'check_and_migrate_config',
+    'config_needs_migrating', 'upgrade_config', 'downgrade_config'
+)
+
+ConfigType = Dict[str, Any]
 
 # The expected version of the configuration file and the oldest backwards compatible configuration version.
 # If the configuration file format is changed, the current version number should be upped and a migration added.
 # When the configuration file format is changed in a backwards-incompatible way, the oldest compatible version should
 # be set to the new current version.
+
 CURRENT_CONFIG_VERSION = 5
 OLDEST_COMPATIBLE_CONFIG_VERSION = 5
 
 
-class ConfigMigration:
-    """Defines a config migration."""
+class SingleMigration(Protocol):
+    """A single migration of the configuration."""
+    initial: int
+    """The initial configuration version."""
+    final: int
+    """The final configuration version."""
+
+    def upgrade(self, config: ConfigType) -> None:
+        """Migrate the configuration in-place."""
 
-    def __init__(self, migrate_function, version, version_oldest_compatible):
-        """Construct a ConfigMigration
+    def downgrade(self, config: ConfigType) -> None:
+        """Downgrade the configuration in-place."""
 
-        :param migrate_function: function which migrates the configuration dictionary
-        :param version: configuration version after the migration.
-        :param version_oldest_compatible: oldest compatible configuration version after the migration.
-        """
-        self.migrate_function = migrate_function
-        self.version = int(version)
-        self.version_oldest_compatible = int(version_oldest_compatible)
 
-    def apply(self, config):
-        """Apply the migration to the configuration."""
-        config = self.migrate_function(config)
-        config.setdefault('CONFIG_VERSION', {})['CURRENT'] = self.version
-        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = self.version_oldest_compatible
-        return config
+class Initial(SingleMigration):
+    """Add the required values for a new default profile.
 
+        * PROFILE_UUID
 
-def _1_add_profile_uuid(config):
+    The profile uuid will be used as a general purpose identifier for the profile, in
+    for example the RabbitMQ message queues and exchanges.
+    """
+    initial = 0
+    final = 1
+
+    def upgrade(self, config: ConfigType) -> None:
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 0
+
+    def downgrade(self, config: ConfigType) -> None:
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 0
```

Comment:
Should the downgrade not _remove_ these keys, since they were added in the upgrade?